### PR TITLE
feat : added utility to read camera live image stream

### PR DESCRIPTION
**Pull request by garuna-m6**

### CODE REVIEW

**Comment by garuna-m6** [comment]
@Davidy22 can you help me in the checks ?

**Comment by dhananjaylatkar** [comment]
> @Davidy22 can you help me in the checks ?

Use `flake8` to see/fix errors.

**Comment by dhananjaylatkar** [comment]
@garuna-m6 which function returns `cv2` or `PIL` image?
I need one of those so that I can feed it into `show_image` function.

**Comment by Trisanu-007** [comment]
@garuna-m6 took care of the checks, enjoy !

**Comment by Davidy22** [comment]
It looks like this does conversion to ASCII straight from openCV which is nice, we need the result as a return value instead of just printed to screen though so that this can be integrated into the rest of the program. Also fold this into a class with cap as an attribute of the class so that it's accessible and we can kill the webcam access.

Also separate the opencv image capture and the conversion to ASCII into two separate functions because @dhananjaylatkar was assigned conversion to ASCII and I believe they have a different-looking converter for the camera output in frame.

File: tox.ini [hunk 1]
```diff
@@ -21,7 +21,9 @@ ignore=
     # Comments
     E266,
     # Type Annotations
-    ANN002,ANN003,ANN101,ANN102,ANN204,ANN206
+    ANN002,ANN003,ANN101,ANN102,ANN204,ANN206,
+    # Some more "S" errors
+    S605,S607
```
Comment (dhananjaylatkar) [inline tox.ini:26]: Instead of ignoring pep8 errors, pls try to fix them.

File: util/CamReader.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+def _open_camera():
+    """
+
+    :return: camera IO object
+    """
+    #  0 -> camera number, if external camera is installed this number needs to be changed
+    # Since a hardware can be only accessible via one user, I/O limitation
+
+    cap = cv.VideoCapture(0)
```
Comment (dhananjaylatkar) [inline util/CamReader.py:20]: this `cap` varriable could go in `__init__` so all methods can access it.

File: util/CamReader.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    # rescaling image to lower dimension
+
+    frame = cv.resize(frame, [w, h])
+    # gray = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
+    # frame = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
+    # Display the resulting frame
+    # frame is the coloured image
+    #  gray is the grayscaled version
+    # gray = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)\
+    return frame
```
Comment (dhananjaylatkar) [inline util/CamReader.py:54]: call `_convert_cv2_to_pil` on `frame` before returning it.

**Review by dhananjaylatkar — CHANGES_REQUESTED** [review]
Add opencv in dev-requirements.txt

File: tox.ini (hunk 1 above)
Comment (Trisanu-007) [reply tox.ini:26]: Tried my best, these errors are happening on the same line and noqa cannot interpret them, leading to fail before commit.
You could replicate this by removing this line and running pre-commit in CamReader.py. The line system("clear") will show the error.

File: util/CamReader.py (hunk 3 above)
Comment (garuna-m6) [reply util/CamReader.py:54]: hmm, better it's called later on numpy image, suppose tomorrow if we want to rollback to numpy image somehow (or use it in the util `method _print_to_dom_color` )

**Review by dhananjaylatkar — CHANGES_REQUESTED** [review]

File: util/CamReader.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+class CamReader():
+    """Utility class to operate camera hardware and capture"""
+
+    def __init__(self):
+        self.cap = self._open_camera()
+
+    @staticmethod
+    def _open_camera() -> cv.VideoCapture:
+        """Opens the camera"""
+        #  0 -> camera number, if external camera is installed this number needs to be changed
```
Comment (dhananjaylatkar) [inline util/CamReader.py:21]: take `0` as default arg. we can ask user to input camera number from our main.py's args

File: util/CamReader.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            print("Can't receive frame (stream end?). Exiting ...")
+        # rescaling image to lower dimension
+        frame = cv.resize(frame, [w, h])
+        # Display the resulting frame
+        # frame is the coloured image
+        #  gray is the grayscaled version
+        # gray = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
+        return frame
+
+    def _convert_cv2_to_pil(self, im: np.ndarray) -> Image.Image:
```
Comment (dhananjaylatkar) [inline util/CamReader.py:56]: `-> Image` is enough

File: tox.ini (hunk 1 above)
Comment (garuna-m6) [reply tox.ini:26]: @dhananjaylatkar can you refer em example for  ANN001 Missing type annotation for function argument 'w'

Comment (dhananjaylatkar) [reply tox.ini:26]: i ran it on your old code. I think you added `int` there. It is fixed now

**Comment by Davidy22** [comment]
Yeah this does what we need it to, can probably come back for cleanup and optimisation but I'll merge for now.

File: util/CamReader.py (hunk 4 above)
Comment (garuna-m6) [reply util/CamReader.py]: hmm, will read about it

File: src/CamReader.py [hunk 6]
```diff
@@ -2,27 +2,14 @@
...
-    def get_time(*args, **kwargs) -> None:
-        """Prints time of the function"""
-        start = time()
-        output = fn(*args, **kwargs)
-        print(f"Time taken in {fn.__name__}: {time() - start:.7f}")
-        return output  # make sure that the decorator returns the output of fn
-
-    return get_time
-
-
```
Comment (garuna-m6) [inline src/CamReader.py]: @Davidy22  We can add this utility in logging

**Pull request merged**